### PR TITLE
2407 custom cache control

### LINKED ISSUE #197: Support custom Cache-Control headers
@davemcorwin commented on [Thu Nov 21 2019](https://github.com/18F/federalist/issues/2407)

 #### User Story

Enable Federalist partners to configure `Cache-Control` headers for their content.

#### Background (Optional)

Federalist partners are currently unable to set custom `Cache-Control` headers for their content and all content is served with the Federalist default.

The current default is `cache-control: max-age=60`, which means that files are only cached for 1 minute by browsers and the CDN.

This is not ideal for our partner's users as they can be leveraging the browser cache and CDN for faster responses for cached requests.

We should enable and add documentation about best practices for caching.

This is also a good time to start implementing a more general configuration file for Federalist configuration instead of piggybacking off of the Jekyll `_config.yml`.

In general, I propose following the APIs ued by [Netlify Configuration File](https://docs.netlify.com/routing/headers/#syntax-for-the-netlify-configuration-file) and [Netlify Headers](https://docs.netlify.com/routing/headers/).

#### Acceptance Criteria

- [ ] Partners can configure `Cache-Control` headers by path
- [ ] Add documentation to Federalist Docs regarding caching best practices
    TLDR:
    - Add unique hash digits to your css, js, and image files (like app-67ce7f3483.css)
    - For js, css, and image files, set `Cache-Control: public, max-age=31536000`
    - For html files, use `Cache-Control: no-cache`

#### Implementation outline (if higher than "low" effort):

- [x] Determine format and name for the federalist configuration file

As json is probably going to be more familiar to our users than toml, I propose a `federalist.json` file that lives in the root of the project.

- [x] Determine format for header configuration
Ex.
```
// federalist.json
// spec
{
  "headers": [
    "<path>": {
      "<header-name>": "<header-value>",
      ...
    },
    ...
  ]
}

// example
{
  "headers": [
    "/*.js": {
      "cache-control": "public, max-age=31536000"
    }
  ]
}
```
- [x] The garden build should read/parse `federalist.json` and have this configuration available
- [x] In the `S3Publisher`, when pushing objects to S3, include any cache-control headers that apply to the object base on its path. This may include including the header information when determining whether an object should be replaced OR refactoring the logic to just replace all files.
- [ ] Potentially add validation of the `federalist.json` file with feedback as part of the build process...

### PR DESCRIPTION
**Pull request by davemcorwin**
Resolves #197 

Herein the ability for a site owner to configure custom `cache-control` header values is added to the Federalist platform.

This PR consists primarily of two parts:
1. A module for handling a `federalist.json` configuration file for the repo with helper methods
2. Updating the publishing logic to add the custom headers provided in `federalist.json` or the default as before.

**NOTE**: Included is a change to the publishing logic that removes the comparison between local and remote files used to limit the number of files uploaded. Instead ALL files are uploaded. This is because the `cache-control` header value would need to be included in this comparison BUT it is not included in the data returned from the call to `list_objects_v2` and would thus require calls to `get_object` for each object. The naming convention in the appropriate places is changed from "modified" to "replacement".

In addition to the unit tests, I ran the application with a sample repository and `federalist.json` file: https://github.com/18F/federalist-uswds-gatsby2/blob/federalist-json/federalist.json. I was able to verify that all files received the correct values for the `cache-control` header.

I would appreciate some extra time reviewing this PR to make sure that the code is correct, idiomatic, well tested, and commented appropriately. In addition, I would like some feedback on the format of the `federalist.json` file, especially given that we have a live example.

Once this is complete, additional issues should be created to add appropriate documentation and/or coordindate an announcement of the feature when ready.

### CODE REVIEW

File: repo_config/repo_config.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    def __init__(self, config={}, defaults={}):
+        self.config = config
+        self.defaults = defaults
+
+    def get_headers_for_path(self, path_to_match):
+        '''
+        Determine the headers that apply to particular file
+        '''
+
+        # Don't think I need a deep copy here
```
Comment (apburnes) [inline repo_config/repo_config.py]: Feel like this is a habit from writing JS. 😆

File: publishing/s3publisher.py [hunk 2]
```diff
@@ -145,30 +179,30 @@ def publish_to_s3(directory, base_url, site_prefix, bucket, cache_control,
...
     ]
 
-    if len(new_objects) == 0 and len(modified_objects) == 0:
+    if len(new_objects) == 0 and len(replacement_objects) == 0:
         raise RuntimeError('Cannot unpublish all files')
 
     LOGGER.info('Preparing to upload')
     LOGGER.info(f'New: {len(new_objects)}')
-    LOGGER.info(f'Modified: {len(modified_objects)}')
+    LOGGER.info(f'Replacement: {len(replacement_objects)}')
```
Comment (amirbey) [inline publishing/s3publisher.py]: I see you renamed the variable from "modified" -> "replaced" which is a naming prefer but perhaps you have good reasoning for changing that?  I think "Modified" is more standard when looking at files ... your thoughts?  I think at least if not for ourselves, the users should see "Modified"

Comment (davemcorwin) [reply publishing/s3publisher.py]: I mentioned this in the diatribe above, but I renamed the variable to be more consistent with what those files actually are now, since I removed the check to determine whether or not they are actually different from the files already on S3. All I know is that they have the same key and will be replaced.

I don't have strong feelings on this, just thought it would be more semantic this way.

File: repo_config/repo_config.py (hunk 1 above)
Comment (davemcorwin) [reply repo_config/repo_config.py]: I specced a bit out in https://github.com/18F/federalist-garden-build/issues/197 and just pushed up some comments to the code that reflect this. One thing to look it is having an explicit spec in the code itself and validating the config against that during the build and throwing errors. I think that would be a nice thing to add.

File: publishing/s3publisher.py (hunk 2 above)
Comment (amirbey) [reply publishing/s3publisher.py]: okay ... to the user can we change from "Replacement" to "Replaced" for consistency
